libjolie/src/main/java/jolie/lang/parse/module/GlobalSymbolReferenceResolver.java

```diff
@@ -0,0 +1,785 @@
+					|| !symbolTables.get( currentURI ).symbol( name ).isPresent() ) {
+					return null;
+				}
+				return symbolTables.get( currentURI ).symbol( name ).get();
+			}, () -> {
+				if( !symbolTables.containsKey( context.source() )
+					|| !symbolTables.get( context.source() ).symbol( name ).isPresent() ) {
+					return null;
+				}
+				return symbolTables.get( context.source() ).symbol( name ).get();
+			} );
+			return symbol;
+		}
+	}
+
+	/**
+	 * perform lookup to external symbol's source of AST node
+	 * 
+	 * @throws SymbolNotFoundException
+	 */
+	private SymbolInfo symbolSourceLookup( SymbolInfoExternal symbolInfo )
+		throws SymbolNotFoundException {
+		ModuleRecord externalSourceRecord =
+			this.moduleMap.get( symbolInfo.moduleSource().get().source() );
+		Optional< SymbolInfo > externalSourceSymbol =
+			externalSourceRecord.symbol( symbolInfo.moduleSymbol() );
+		if( !externalSourceSymbol.isPresent() ) {
+			throw new SymbolNotFoundException( symbolInfo.name(), symbolInfo.moduleTargets() );
+		}
+		if( externalSourceSymbol.get().scope() == Scope.LOCAL ) {
+			return externalSourceSymbol.get();
+		} else {
+			return symbolSourceLookup( (SymbolInfoExternal) externalSourceSymbol.get() );
+		}
+	}
+
+	/**
+	 * Find and set a pointer of externalSymbol to it's corresponding AST node by perform lookup at
+	 * ModuleRecord Map, a result from ModuleCrawler.
+	 * 
+	 * @throws DuplicateSymbolException
+	 * @throws IllegalAccessSymbolException
+	 * @throws SymbolNotFoundException
+	 * 
+	 */
+	public void resolveExternalSymbols()
+		throws SymbolNotFoundException, IllegalAccessSymbolException, DuplicateSymbolException {
+		for( ModuleRecord md : moduleMap.values() ) {
+			for( SymbolInfoExternal localSymbol : md.externalSymbols() ) {
+				if( localSymbol instanceof SymbolWildCard ) {
+					ModuleRecord wildcardImportedRecord =
+						this.moduleMap.get( localSymbol.moduleSource().get().source() );
+					md.addWildcardImportedRecord( (SymbolWildCard) localSymbol,
+						wildcardImportedRecord.symbols() );
+				} else {
+					SymbolInfo targetSymbol = symbolSourceLookup( localSymbol );
+					if( targetSymbol.privacy() == Privacy.PRIVATE ) {
+						throw new IllegalAccessSymbolException( localSymbol.name(),
+							localSymbol.moduleTargets() );
+					}
+					localSymbol.setPointer( targetSymbol.node() );
+				}
+			}
+		}
+	}
+
+	/**
+	 * resolve LinkedType of each ModuleRecord AST node in the Map.
+	 * 
+	 * @throws ModuleException if the linked type cannot find it's referencing node
+	 */
+	public void resolveLinkedType() throws ModuleException {
+		SymbolReferenceResolverVisitor resolver = new SymbolReferenceResolverVisitor();
+		for( ModuleRecord md : moduleMap.values() ) {
+			resolver.resolve( md.program() );
+		}
+	}
+
+	/**
+	 * Resolve symbols the is imported from external modules and resolve linked type's pointer
+	 * 
+	 * @throws ModuleException if the process is failed
+	 */
+	public void resolve() throws ModuleException {
```

Comment:
Could be a static method (you always create the resolver right before).

Also, GlobalSymbolReferenceResolver could just be SymbolReferenceResolver? :-)